### PR TITLE
Migration to Python 3.7+ including boto -> boto3

### PR DESCRIPTION
**Pull request by lod**
Basic testing done, passes all unit tests locally and docker container is created successfully

Ready for high level review

### CODE REVIEW

**Comment by lod** [comment]
I intend to run black to reformat and hide my uglys

But it will completely obscure what was changed so I've held off for now

**Comment by lod** [comment]
These were flagged in the code, but pulling them out so that they don't get lost
Some things to ensure are tested

* Need to check partial transfer behaviour, ensure we clean up
* Ensure we can push without list permissions

File: ingester/tests/test_ingester.py [hunk 1]
```diff
@@ -22,66 +25,68 @@ def maker():
     return maker
 
 
+@mock.patch.dict(os.environ, {"AWS_REGION":"us-east-1"})
```
Comment (bcavagnolo) [inline ingester/tests/test_ingester.py:28]: Can you use pytest's `monkeypatch` facility instead? Should be `monkeypatch.setenv("AWS_REGION", "us-east-1")`

File: ingester/tests/test_ingester.py [hunk 2]
```diff
@@ -22,66 +25,68 @@ def maker():
...
-    records = [dict(r) for r in dynamodb_records_table.scan()]
-    assert len(records) >= 1
-    for r in records:
+    scan_r = dynamodb_records_table.scan()
+    assert scan_r['Count'] >= 1
+    for r in scan_r['Items']:
         assert r['metadata'] == metadata
 
 
+@mock.patch.dict(os.environ, {"AWS_REGION":"us-east-1"})
```
Comment (bcavagnolo) [inline ingester/tests/test_ingester.py:39]: ...and here too.

File: client/test/conftest.py [hunk 3]
```diff
@@ -69,23 +64,24 @@ def archive(archive_maker):
...
-            keys = [k for k in s3_bucket.list()]
-            assert len(keys) == 1
-            return keys[0]
+            objs = s3_bucket.objects.all()
+            assert len(objs) == 1
+            return objs[0].Object()
         else:
+            print("A1", url)
             url = urlparse(url)
+            print("A2", url)
```
[the remark this answers is not in the document]
Comment (lod) [reply client/test/conftest.py]: Fixed in
[e98f2ae](https://github.com/planetlabs/datalake/pull/78/commits/e98f2aee87e2302a955ac676d46536e3b3fcc476)

File: client/test/test_file.py [hunk 4]
```diff
@@ -42,6 +42,9 @@ def get_randfiles(n):
 
 def test_file_hash_different(random_files):
     files = random_files(2)
+    print(files[0])
+    print(files[1])
+
```
[the remark this answers is not in the document]
Comment (lod) [reply client/test/test_file.py:46]: Fixed in
[e98f2ae](https://github.com/planetlabs/datalake/pull/78/commits/e98f2aee87e2302a955ac676d46536e3b3fcc476)

File: client/datalake/archive.py [hunk 5]
```diff
@@ -357,18 +364,36 @@ def _mkdirs(self, path):
...
-    def _get_key_from_url(self, url):
-        self._validate_fetch_url(url)
-        key_name = self._get_key_name_from_url(url)
-        k = self._s3_bucket.get_key(key_name)
-        if k is None:
-            msg = 'Failed to find {} in the datalake.'.format(url)
-            raise InvalidDatalakePath(msg)
-        return k
+    @staticmethod
+    def split_s3_url(url):
```
[the remark this answers is not in the document]
Comment (lod) [reply client/datalake/archive.py]: Fixed in [6e8bc7d](https://github.com/planetlabs/datalake/pull/78/commits/6e8bc7d6183ed1b3032ec39fc3291ac5de35bb9e)

File: .github/workflows/actions.yml [hunk 6]
```diff
@@ -7,7 +7,7 @@ jobs:
     strategy:
       matrix:
         # 4 jobs total with all permutations
-        python: [2.7, 3.5]
+        python: [2.7, 3.10.1]
```
[the remark this answers is not in the document]
Comment (lod) [reply .github/workflows/actions.yml]: fixed in [449c16d](https://github.com/planetlabs/datalake/pull/78/commits/449c16d36f35e436455b15609861264b825af0a7)

File: ingester/tests/test_ingester.py (hunk 1 above)
Comment (lod) [reply ingester/tests/test_ingester.py]: Thanks, didn't know about monkeypatch

Fixed in [58ceef1](https://github.com/planetlabs/datalake/pull/78/commits/58ceef10364b8b918c670b84da66bbd9882d3175)

**Comment by lod** [comment]
An update as we approach the end of the week.

I've addressed most of the issues raised. Mostly the easier ones.  
The others will be addressed, soon.

Updated the api project from py2 to py3. Required because it uses the client which is now py3, and the container which is now py3. Had previously not done it because it already used boto3 and I naively assumed all was swell.

Some commits to fix up tests which I missed somehow, and tests that were skipped.

Also fixed up the github unit test automation. Now it covers off all the currently supported python releases and all three of the datalake projects. Also made the tests actually pass.

Next big block of work is testing the client upload flagged special cases. Pushing with very constrained permissions, and multipart upload failures. Will also try and expand the unit tests to cover these.

File: Dockerfile [hunk 7]
```diff
@@ -51,7 +26,6 @@ RUN for d in client ingester api; do \
     cd /opt/$d && \
     python setup.py develop -s /usr/local/bin \
         --egg-path ../../../../../opt/$d/ \
-        -d /usr/local/lib/python2.7/site-packages/ \
```
[the remark this answers is not in the document]
Comment (lod) [reply Dockerfile:54]: Fixed in 9e420868502e9d4e3e99e00f7c7fc7d6b26b39f2

setup.py develop is actually depreciated.

So I migrated to pip -e . which does what is required (tested) and avoids the python version path

**Comment by lod** [comment]
I did some testing of the failed multipart cleanup behaviour, specifically I did a multipart upload and then killed my network link.

The partial upload is still sitting there. Which makes sense, it's impossible to cleanup when you don't have access to do so.

I suggest we implement a lifecycle rule as described at https://aws.amazon.com/blogs/aws-cloud-financial-management/discovering-and-deleting-incomplete-multipart-uploads-to-lower-amazon-s3-costs/

Govcloud doesn't seem to support the S3 lens for analysis. Partial uploads can be found using
`aws s3api list-multipart-uploads --bucket datalake-staging`

There is currently 1 partial in staging (mine), and 477 in prod.

**Pull request closed without merging**